src/annotator/util/buckets.js

```diff
@@ -0,0 +1,248 @@
+ * @return {PositionPoints}
+ */
+export function constructPositionPoints(anchors) {
+  const aboveScreenAnchors = new Set();
+  const belowScreenAnchors = new Set();
+  const points = /** @type {PositionPoint[]} */ (new Array());
+
+  for (let anchor of anchors) {
+    if (!anchor.highlights?.length) {
+      continue;
+    }
+
+    const rect = getBoundingClientRect(anchor.highlights);
+
+    if (rect.top < BUCKET_TOP_THRESHOLD) {
+      aboveScreenAnchors.add(anchor);
+    } else if (rect.top > window.innerHeight - BUCKET_NAV_SIZE) {
+      belowScreenAnchors.add(anchor);
+    } else {
+      // Add a point for the top of this anchor's highlight box
+      points.push([rect.top, 1, anchor]);
+      // Add a point for the bottom of this anchor's highlight box
+      points.push([rect.bottom, -1, anchor]);
+    }
+  }
+
+  // Sort onscreen points by pixel position, secondarily by position "type"
+  // (top or bottom of higlight box)
+  points.sort((a, b) => {
+    for (let i = 0; i < a.length; i++) {
+      if (a[i] < b[i]) {
+        return -1;
+      } else if (a[i] > b[i]) {
+        return 1;
+      }
+    }
+    return 0;
+  });
+
+  return {
+    above: Array.from(aboveScreenAnchors),
+    below: Array.from(belowScreenAnchors),
+    points,
+  };
+}
+
+/**
+ * Take a sorted set of `points` representing top and bottom positions of anchor
+ * highlights and group them into a collection of "buckets".
+ *
+ * @param {PositionPoint[]} points
+ * @return {BucketInfo}
+ */
+export function buildBuckets(points) {
+  const buckets = /** @type {Array<Anchor[]>} */ (new Array());
+  const bucketPositions = /** @type {number[]} */ (new Array());
+
+  // Anchors that are part of the currently-being-built bucket, and a correspon-
+  // ding count of unclosed top edges seen for that anchor
+  const current = /** @type {{anchors: Anchor[], counts: number[] }} */ ({
+    anchors: new Array(),
+    counts: new Array(),
+  });
+
+  points.forEach((point, index) => {
+    const [position, delta, anchor] = point;
+    // Does this point represent the top or the bottom of an anchor's highlight
+    // box?
+    const positionType = delta > 0 ? 'start' : 'end';
+
+    // See if this point's anchor is already in our working set of open anchors
+    const anchorIndex = current.anchors.indexOf(anchor);
+
+    if (positionType === 'start') {
+      if (anchorIndex === -1) {
+        // Add an entry for this anchor to our current set of "open" anchors
+        current.anchors.unshift(anchor);
+        current.counts.unshift(1);
+      } else {
+        // Increment the number of times we've seen a start/top edge for this
+        // anchor
+        current.counts[anchorIndex]++;
+      }
+    } else {
+      // positionType = 'end'
+      // This is the bottom/end of an anchor that we should have already seen
+      // a top edge for. Decrement the count, representing that we've found an
+      // end point to balance a previously-seen start point
```

Comment:
Some of this is pure inference as to what I believe the original algorithm was attempting to do. I'm confident the answers the migrated variant get are correct; I'm less entirely confident that my explanation is accurate to the original author's intent.